src/category_theory/monad/equiv_mon.lean

```diff
@@ -0,0 +1,233 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+import category_theory.monad.bundled
+import category_theory.monoidal.End
+import category_theory.monoidal.internal
+import category_theory.category.Cat
+
+/-!
+
+# The equivalence between `Monad C` and `Mon_ (C ⥤ C)`.
+
+A monad "is just" a monoid in the category of endofunctors.
+
+# Definitions/Theorems
+
+1. `to_Mon` associates a monoid object in `C ⥤ C` to any monad on `C`.
+2. `Monad_to_Mon` is the functorial version of `to_Mon`.
+3. `of_Mon` associates a monad on `C` to any monoid object in `C ⥤ C`.
+4. `Mon_to_Monad` is the functorial version of `of_Mon`.
+5. `Monad_Mon_equiv` is the equivalence between `Monad C` and `Mon_ (C ⥤ C)`.
+  NB: It is really an isomorphism of categories!
+
+The primary purpose for the theorems in this file is to construct `Monad_Mon_equiv`.
+1. `of_to_mon_end_obj` and `to_of_mon_end_obj` show that the compositions of
+  `Mon_to_Monad` and `Monad_to_Mon` act as the identity on objects.
+2. `of_to_mon_end` and `to_of_mon_end` promote the equalities from these two theorems to
+  equalities with the identity functor.
+
+# Note
+
+It is tempting to use `eq_to_iso` along with `of_to_mon_end` and `to_of_mon_end` in order to
+construct `Monad_Mon_equiv`, or equivalently, to use `Cat.equiv_of_iso` applied to `Monad_Mon_iso`.
+Unfortunately, such an equivalence is difficult to use in practice.
+
+Because of this, we construct `Monad_Mon_equiv` "by hand" using `of_to_mon_end_iso` and
+`to_of_mon_end_iso`. See the "sanity check" example in this file.
+
+-/
+
+namespace category_theory
+open category
+
+universes v u -- declare the `v`'s first; see `category_theory.category` for an explanation
+variables {C : Type u} [category.{v} C]
+
+namespace Monad
+local attribute [instance] endofunctor_monoidal_category
+
+/-- To every `Monad C` we associated a monoid object in `C ⥤ C`.-/
+def to_Mon : Monad C → Mon_ (C ⥤ C) := λ M,
+{ X := M.func,
+  one := η_ _,
+  mul := μ_ _,
+  one_mul' := begin
+    change (_ ◫ _) ≫ _ = _,
+    ext A,
+    simp only [nat_trans.hcomp_id_app, nat_trans.comp_app],
+    apply monad.right_unit,
+  end,
+  mul_one' := begin
+    change (_ ◫ _) ≫ _ = _,
+    tidy,
+  end,
+  mul_assoc' := begin
+    change (_ ◫ _) ≫ _ = _ ≫ (_ ◫ _) ≫ _,
+    ext A,
+    simp only [nat_trans.hcomp_id_app, nat_trans.hcomp_app, functor.map_id,
+      nat_trans.id_app, comp_id, nat_trans.comp_app],
+    erw id_comp,
+    simp_rw monad.assoc,
+    change _ = ((α_ M.func M.func M.func).app A).hom ≫ _ ≫ _,
+    suffices : ((α_ M.func M.func M.func).app A).hom = 𝟙 _, by {rw this, simp},
+    refl,
+  end }
+
+variable (C)
+/-- Passing from `Monad C` to `Mon_ (C ⥤ C)` is functorial. -/
+def Monad_to_Mon : Monad C ⥤ Mon_ (C ⥤ C) :=
+{ obj := to_Mon,
+  map := λ M N f,
+  { hom := f.to_nat_trans,
+    one_hom' := begin
+      ext,
+      simp only [nat_trans.comp_app],
+      apply f.app_η,
```

Comment:
This is a good example of something that I'd prefer worked by `simp`.

Reply:
(and in fact could even then be omitted because an `auto_param` would handle it)